api/classes/class-lightweight-api.php

```diff
@@ -0,0 +1,303 @@
+		$this->has_newsletter_prompt = ! empty( $_REQUEST['has_newsletter_prompt'] ) ? // phpcs:ignore
+			filter_input( INPUT_POST | INPUT_GET, 'has_newsletter_prompt', FILTER_SANITIZE_SPECIAL_CHARS ) :
+			null; // phpcs:ignore
+
+		$this->referer_url = filter_input( INPUT_SERVER, 'HTTP_REFERER', FILTER_SANITIZE_STRING );
+
+		if ( 'inline' !== $this->placement && 'always' === $this->frequency ) {
+			$this->frequency = 'once';
+		}
+	}
+
+	/**
+	 * Verify referer is valid.
+	 */
+	public function verify_referer() {
+		$http_referer = ! empty( $_SERVER['HTTP_REFERER'] ) ? parse_url( $_SERVER['HTTP_REFERER'] , PHP_URL_HOST ) : null; // phpcs:ignore
+		$valid_referers = [
+			$http_referer,
+			// TODO: Add AMP Cache.
+		];
+		$http_host = ! empty( $_SERVER['HTTP_HOST'] ) ? $_SERVER['HTTP_HOST'] : null; // phpcs:ignore
+		return in_array( strtolower( $http_host ), $valid_referers, true );
+	}
+
+	/**
+	 * Establish DB connection.
+	 */
+	public function connect() {
+		$credentials = $this->get_credentials();
+		if ( ! $credentials ) {
+			$this->error( 'no_credentials' );
+		}
+		$db = mysqli_init(); //phpcs:ignore
+		$db->real_connect(
+			$credentials['db_host'],
+			$credentials['db_user'],
+			$credentials['db_password'],
+			$credentials['db_name']
+		);
+		return $db;
+	}
+
+	/**
+	 * Get database credentials from environment variables (Atomic) or special config file.
+	 */
+	public function get_credentials() {
+		if ( getenv( 'DB_NAME' ) && getenv( 'DB_USER' ) && getenv( 'DB_PASSWORD' ) ) {
+			return [
+				'db_host'     => 'localhost',
+				'db_name'     => getenv( 'DB_NAME' ),
+				'db_user'     => getenv( 'DB_USER' ),
+				'db_password' => getenv( 'DB_PASSWORD' ),
+			];
+		}
+		$config = $_SERVER['DOCUMENT_ROOT'] . '/newspack-popups-config.php'; //phpcs:ignore
+		if ( file_exists( $config ) ) {
+			require $config;
+			return [
+				'db_host'     => DB_HOST,
+				'db_name'     => DB_NAME,
+				'db_user'     => DB_USER,
+				'db_password' => DB_PASSWORD,
+			];
+		}
+		return null;
+	}
+
+	/**
+	 * Get transient name.
+	 */
+	public function transient_name() {
+		return sprintf( '_transient_%s-%s-popup', $this->rid, $this->popup_id );
+	}
+
+	/**
+	 * Get suppression-related transient value
+	 *
+	 * @param string $prefix transient prefix.
+	 */
+	public function get_suppression_data_transient_name( $prefix ) {
+		return sprintf(
+			'_transient_%s-%s',
+			$prefix,
+			$this->rid
+		);
+	}
+
+	/**
+	 * Get data from transient.
+	 *
+	 * @param string $transient_name The transient's name.
+	 */
+	public function get_transient( $transient_name ) {
+		$query                    = sprintf(
+			"SELECT option_value FROM wp_options WHERE option_name = '%s'",
+			$this->db->real_escape_string( $transient_name )
+		);
+		$this->debug['queries'][] = $query;
+		$result                   = $this->db->query( $query );
+		$row                      = $result->fetch_assoc();
+		$transient                = maybe_unserialize( $row['option_value'] );
```

Comment:
This should check to make sure something was actually returned from the DB before referencing `$row['option_value']`

Reply:
Done in https://github.com/Automattic/newspack-popups/pull/208/commits/aac241f2d83fc5688b32f5d286fa5b4931fb91a7